### DISCUSSION TITLE
Primer React v36 Release

### DISCUSSION THREAD

**Discussion by joshblack** [primer/react#3889] [Announcements]
We're excited to announce the next major release of Primer React! 🥳 Starting in v36, you'll see changes to `@primer/react` that make it even easier to build consistent, accessible experiences that feel right at home at GitHub.

**We'd love to hear from you and get feedback on these changes**. Let us know in a comment below if you run into any bugs, if a change to a component you were using. is unclear, or if you have any other feedback about this release.

## New components :sparkles:

The following components have been moved from `@primer/react/drafts` to `@primer/react`:

- [`TreeView`](https://primer.style/design/components/tree-view)
- [`NavList`](https://primer.style/design/components/nav-list)
- [`SegmentedControl`](https://primer.style/design/components/segmented-control)
- [`SplitPageLayout`](https://primer.style/design/components/split-page-layout/react)

## Breaking changes ⚠️

The v36 release of `@primer/react` introduces several breaking changes to components and changes to the package itself. To understand how to migrate between v35 and v36 of `@primer/react`, review the sections below to understand how your project may be impacted by these changes.

### Components

The following list of components have breaking changes to their API or availability in `@primer/react`. For a list of deprecated components that have been removed, visit the [Removed components](#removed-components) section below.

- [`AvatarToken`](#avatartoken)
- [`Button`](#button)
- [`Checkbox`](#checkbox)
- [`FilterList`](#filterlist)
- [`FilterSearch`](#filtersearch)
- [`FormControl`](#formcontrol)
- [`Radio`](#radio)
- [`Textarea`](#textarea)
- [`TextInputWithTokens`](#textinputwithtokens)
- [`UnderlineNav`](#underlinenav)

#### AvatarToken

The `size` prop for `AvatarToken` has been updated in v36 to only accept `"large"` and `"xlarge"`.

<table>
<thead>
<tr>
  <th>v35</th>
  <th>v36</th>
</tr>
</thead>
<tbody>
<tr>
<td valign="top">
        
```jsx
<AvatarToken size="small">
<AvatarToken size="medium">
<AvatarToken size="large">
<AvatarToken size="xlarge">
<AvatarToken size="extralarge">
```
        
</td>
<td valign="top">
        
```jsx
<AvatarToken size="large">
<AvatarToken size="xlarge">
```
  
</td>
</tr>
</tbody>
</table>

#### Button

The API for the `Button` component has been updated to match the specification for this component and align with other implementations like in Primer View Components.

- The `leadingIcon` prop has been renamed to `leadingVisual`

    <table>
    <thead>
    <tr>
      <th>v35</th>
      <th>v36</th>
    </tr>
    </thead>
    <tbody>
    <tr>
    <td valign="top">
            
    ```jsx
    <Button leadingIcon={TrashIcon}>Delete</Button>
    ```
            
    </td>
    <td valign="top">
            
    ```jsx
    <Button leadingVisual={TrashIcon}>Delete</Button>
    ```
      
    </td>
    </tr>
    </tbody>
    </table>

- The `trailingIcon` prop has been renamed to `trailingVisual`

    <table>
    <thead>
    <tr>
      <th>v35</th>
      <th>v36</th>
    </tr>
    </thead>
    <tbody>
    <tr>
    <td valign="top">
            
    ```jsx
    <Button trailingIcon={EyeIcon}>Delete</Button>
    ```
            
    </td>
    <td valign="top">
            
    ```jsx
    <Button trailingVisual={EyeIcon}>Delete</Button>
    ```
      
    </td>
    </tr>
    </tbody>
    </table>
    
- The `Button.Counter` component has been removed, use the `count` prop instead

    <table>
    <thead>
    <tr>
      <th>v35</th>
      <th>v36</th>
    </tr>
    </thead>
    <tbody>
    <tr>
    <td valign="top">
            
    ```jsx
    <Button>
      Watch
      <Button.Counter>5</Button.Counter>
    </Button>
    ```
            
    </td>
    <td valign="top">
            
    ```jsx
    <Button count={5}>Watch</Button>
    ```
      
    </td>
    </tr>
    </tbody>
    </table>

- The `outline` variant has been removed, instead use `variant="invisible"`

    <table>
    <thead>
    <tr>
      <th>v35</th>
      <th>v36</th>
    </tr>
    </thead>
    <tbody>
    <tr>
    <td valign="top">
            
    ```jsx
    <Button variant="outline">Example</Button>
    ```
            
    </td>
    <td valign="top">
            
    ```jsx
    <Button variant="invisible">Example</Button>
    ```
      
    </td>
    </tr>
    </tbody>
    </table>

#### Checkbox
  
The `validationStatus` prop has been updated and no longer supports the `'warning'` value. Instead, only `'error'` and `'success'` are supported.
  
#### FilterList

The `FilterList` component has been deprecated in v36 in favor of the [`ActionList`](https://primer.style/react/ActionList) component. To continue using this component, update your imports for `FilterList` to import from the `deprecated` entrypoint.

<table>
<thead>
<tr>
  <th>v35</th>
  <th>v36</th>
</tr>
</thead>
<tbody>
<tr>
<td valign="top">
        
```jsx
import { FilterList } from '@primer/react';
```
        
</td>
<td valign="top">
        
```jsx
import { FilterList } from '@primer/react/deprecated';
```
  
</td>
</tr>
</tbody>
</table>

#### FilterSearch

The `FilterSearch` component has been deprecated in v36. A  new filter component is currently in progress but is not yet available. Until a new component is ready, use a combination of [`Button`](https://primer.style/react/Button), [`TextInput`](https://primer.style/react/TextInput), and [`ActionList`](https://primer.style/react/ActionList) to reproduce this component. To continue using this component, update your imports for `FilterSearch` to import from the `deprecated` entrypoint.

<table>
<thead>
<tr>
  <th>v35</th>
  <th>v36</th>
</tr>
</thead>
<tbody>
<tr>
<td valign="top">
        
```jsx
import { FilterSearch } from '@primer/react';
```
        
</td>
<td valign="top">
        
```jsx
import { FilterSearch } from '@primer/react/deprecated';
```
  
</td>
</tr>
</tbody>
</table>

#### FormControl
  
The `variant` prop for the `FormControl.Validation` component has been updated to no longer support the `'warning'` value. Instead, only `'error'` and `'success'` are supported.
    
#### Radio
  
The `validationStatus` prop has been updated and no longer supports the `'warning'` value. Instead, only `'error'` and `'success'` are supported.

#### Textarea
  
The `validationStatus` prop has been updated and no longer supports the `'warning'` value. Instead, only `'error'` and `'success'` are supported.

#### TextInputWithTokens
    
The `size` prop has been updated to no longer accept the `extralarge` value. Use `xlarge` instead.
    
<table>
<thead>
<tr>
  <th>v35</th>
  <th>v36</th>
</tr>
</thead>
<tbody>
<tr>
<td valign="top">
        
```jsx
<TextInputWithTokens size="extralarge">
```
        
</td>
<td valign="top">
        
```jsx
<TextInputWithTokens size="xlarge">
```
  
</td>
</tr>
</tbody>
</table>

#### UnderlineNav

The implementation for `UnderlineNav` has been updated in v36. Certain `props` for UnderlineNav have been removed and the `UnderlineNav.Link` component has been replaced by `UnderlineNav.Item`.

| Component | v35 | v36 |
| :-------- | :-- | :-- |
| UnderlineNav | `actions` | This prop has been removed |
|              | `align` | This prop has been removed |
|              | `full` | This prop has been removed |
| UnderlineNav.Link | | Use the `UnderlineNav.Item` component |
| UnderlineNav.Link | `selected` | Use the `aria-current` prop |

<table>
<thead>
<tr>
  <th>v35</th>
  <th>v36</th>
</tr>
</thead>
<tbody>
<tr>
<td valign="top">
        
```jsx
<UnderlineNav aria-label="Main">
  <UnderlineNav.Link href="#home" selected>
    Home
  </UnderlineNav.Link>
  <UnderlineNav.Link href="#documentation">
    Documentation
  </UnderlineNav.Link>
  <UnderlineNav.Link href="#support">
    Support
  </UnderlineNav.Link>
</UnderlineNav>
```
        
</td>
<td valign="top">
        
```jsx
<UnderlineNav aria-label="Main">
  <UnderlineNav.Item href="#home" aria-current="page">
    Home
  </UnderlineNav.Item>
  <UnderlineNav.Item href="#documentation">
    Documentation
  </UnderlineNav.Item>
  <UnderlineNav.Item href="#support">
    Support
  </UnderlineNav.Item>
</UnderlineNav>
```
  
</td>
</tr>
</tbody>
</table>

To continue using the implementation for `UnderlineNav` from `v35`, import the component from `@primer/react/deprecated`.

<table>
<thead>
<tr>
  <th>v35</th>
  <th>v36</th>
</tr>
</thead>
<tbody>
<tr>
<td valign="top">
        
```jsx
import { UnderlineNav } from '@primer/react';
```
        
</td>
<td valign="top">
        
```jsx
import { UnderlineNav } from '@primer/react/deprecated';
```
  
</td>
</tr>
</tbody>
</table>

#### Removed components

The following deprecated components have been removed from `@primer/react` and are no longer available.

| Name | Migration |
| :--- | :-------- |
| `BorderBox` | Use the Box component, `<BorderBox>` → `<Box borderWidth='1px' borderStyle='solid' borderColor='border.primary' borderRadius={2}>` |
| `ChoiceFieldset` | Use the [`CheckboxGroup`](https://primer.style/react/CheckboxGroup) or [`RadioGroup`](https://primer.style/react/RadioGroup) component |
| `Item` | Use [`CheckboxGroup`](https://primer.style/react/CheckboxGroup) or [`RadioGroup`](https://primer.style/react/RadioGroup) component |
| `ChoiceInputField` | Use the [`FormControl`](https://primer.style/react/FormControl) component |
| `Flex` | Use the [`Box`](https://primer.style/design/components/box) component, `<Flex>` → `<Box display="flex">` |
| `Grid` | Use the [`Box`](https://primer.style/design/components/box) component, `<Grid>` → `<Box display="grid">` |
| `Position` | Use the [`Box`](https://primer.style/design/components/box) component, `<Position>` → `<Box>` |
| `Absolute` | Use the [`Box`](https://primer.style/design/components/box) component, `<Absolute>` → `<Box position="absolute">` |
| `Fixed` | Use the [`Box`](https://primer.style/design/components/box) component, `<Fixed>` → `<Box position="fixed">` |
| `Relative` | Use the [`Box`](https://primer.style/design/components/box) component, `<Relative>` → `<Box position="relative">` |
| `Sticky` | Use the [`Box`](https://primer.style/design/components/box) component, `<Sticky>` → `<Box position="sticky">` |
| `Dropdown` | Use the [`ActionMenu`](https://primer.style/react/ActionMenu) component |
| `FormGroup` | Use the [`FormControl`](https://primer.style/react/FormControl) component |
| `InputField` | Use the [`FormControl`](https://primer.style/react/FormControl) component |
| `Label` | Use the updated [`Label`](https://primer.style/design/components/label/react) component from the `@primer/react` entry point |
| `SelectMenu` | Use the [`ActionMenu`](https://primer.style/react/ActionMenu) component |
| `ActionList` | Use the updated [`ActionList`](https://primer.style/react/ActionList) component from the `@primer/react` entry point |
| `ActionMenu` | Use the updated [`ActionMenu`](https://primer.style/react/ActionMenu) component from the `@primer/react` entry point |
| `DropdownButton` | Use the [`ActionMenu`](https://primer.style/react/ActionMenu) component |
| `DropdownMenu` | Use the [`ActionMenu`](https://primer.style/react/ActionMenu) component |
| `Button` | Use the updated [`Button`](https://primer.style/react/Button) component from the `@primer/react` entry point |
| `ButtonDanger` | Use the updated [`Button`](https://primer.style/react/Button) component with `variant="danger"` |
| `ButtonOutline` | Use the updated [`Button`](https://primer.style/react/Button) component with `variant="outline"` |
| `ButtonPrimary` | Use the updated [`Button`](https://primer.style/react/Button) component with `variant="primary"` |
| `ButtonInvisible` | Use the updated [`Button`](https://primer.style/react/Button) component with `variant="invisible"` |
| `ButtonTableList` | Use the [`ActionMenu`](https://primer.style/react/ActionMenu) component |
| `ButtonClose` | Use the [`IconButton`](https://primer.style/react/IconButton) component, `<IconButton icon={XIcon} aria-label="Close" />` |
| `StyledOcticon` | Use the [`Octicon`](https://primer.style/react/Octicon) component |

#### Changes to peer dependencies 🏗️ 

The following peer dependencies have been updated, or added, to require a new minimum version in v36 of `@primer/react`:

| Package | v35 | v36 |
| :------ | :-- | :-- |
| `react` | `^17.0.0 \|\| ^18.0.0` | `^18.0.0` |
| `react-dom` | `^17.0.0 \|\| ^18.0.0` | `^18.0.0` |
| `styled-components` | `4.x \|\| 5.x` | `5.x` |
| `@types/styled-components` | Not available | `5.x` |

These dependencies will need to be updated or installed in your project if you are not using the corresponding version range for v36. To see if you're using the correct versions, look in your project's `package.json` file and see if the versions match the table above.

> **Note**
> If you're not using TypeScript in your project, feel free to ignore the `@types` dependencies above.

#### `drafts` has been moved to `experimental`

The `drafts` entrypoint has been deprecated in v35 and removed in v36. It has been replaced by the `experimental` entrypoint to better reflect the [component lifecycle](https://primer.style/design/guides/component-lifecycle/) process for Primer.

<table>
<thead>
<tr>
  <th>v35</th>
  <th>v36</th>
</tr>
</thead>
<tbody>
<tr>
<td valign="top">

```jsx
import { ExampleComponent } from '@primer/react/drafts';
```

</td>
<td valign="top">

```jsx
import { ExampleComponent } from '@primer/react/experimental';
```

</td>
</tr>
</tbody>
</table>

**Comment by gr2m** [comment]
is it possible that `StyledOcticon` was removed in v36? It's not listed in the breaking changes

**Reply by joshblack** [reply]
Hey @gr2m! You're right, that was totally not included in the changelog (I'll add it in).

StyledOcticon is now just Octicon 👍
